### PR TITLE
feat: Interaction components

### PR DESCRIPTION
**Pull request by mikhail-dcl**
Two different kinds of raycast processing have been introduced:

- Player Originated
   - Responsible for Input and Raycast Propagation
   - `PBPointerEvents` and `PBPointerEventsResult` SDK components
   - Global Input Processing which is unified with these events (very strange concept 😢 )
- Entity Originated
   - `PBRaycast` and `PBRaycastResult` components
   - Budgeted
- Hover Canvas based on UIToolkit
- Collider to Entity cache both in the scene and global context 
- Had to refactor `ECSToCRDTWriter` as it was never tested before and exposed some critical issues
- Added Camera Component propagation as an experiment (not very successful after all, requires more time)

Existing issues:
- Sometimes Input is ignored by the scene, most likely some concurrency issues. Average severity
- `ECSToCRDTWriter` shows bad performance, not scalable as it's executed from the main thread. To be addressed in the next shape-up. Most likely we will need to jobify it
- Concurrency issues on writing LWW components (was out of scope)
- Considerable allocations from **ClearScript** side

### CODE REVIEW

File: Explorer/Assets/DCL/Interaction/PlayerOriginated/Systems/ProcessPointerEventsSystem.cs [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+            if (!stateComponent.LastHitCollider) // collider was destroyed, nothing to do
+            {
+                globalColliderEntityInfo = default(GlobalColliderEntityInfo);
+                return false;
+            }
+
+            return entityCollidersGlobalCache.TryGetEntity(stateComponent.LastHitCollider, out globalColliderEntityInfo); // scene was destroyed, collider was returned to the pool, nothing to do
+        }
+
+        [Query]
```
Comment (popuz) [inline Explorer/Assets/DCL/Interaction/PlayerOriginated/Systems/ProcessPointerEventsSystem.cs:47]: this method is huge and contains several nesting levels. It leads to high cognitive complexity. I suggest to split it into readable sub-functions

File: Explorer/Assets/DCL/Interaction/HoverCanvas/UI/HoverCanvas.cs [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+            if (initialized) return;
+
+            tooltips = this.Query<HoverCanvasTooltipElement>().ToList();
+            tooltips.Sort(this);
+            initialized = true;
+        }
+
+        public void SetTooltip([CanBeNull] string hintText, [CanBeNull] string actionKeyText, [CanBeNull] Sprite icon, int index)
+        {
+            if (index >= tooltips.Count)
```
[the remark this answers is not in the document]
Comment (mikhail-dcl) [reply Explorer/Assets/DCL/Interaction/HoverCanvas/UI/HoverCanvas.cs:26]: It's a content-creator mistake, I don't think that we have an agreed way of reporting them

File: Explorer/Assets/DCL/Interaction/Utility/RaycastUtils.cs [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+            Vector3 rayOrigin = entityTransform.Cached.WorldPosition + sdkRaycast.OriginOffset;
+            Vector3 rayDirection;
+
+            switch (sdkRaycast.DirectionCase)
+            {
+                case PBRaycast.DirectionOneofCase.LocalDirection:
+                    rayDirection = entityTransform.Cached.WorldRotation * sdkRaycast.LocalDirection;
+                    break;
+                case PBRaycast.DirectionOneofCase.GlobalTarget:
+                    rayDirection = sceneRootPos + sdkRaycast.GlobalTarget - entityTransform.Cached.WorldPosition;
```
[the remark this answers is not in the document]
Comment (mikhail-dcl) [reply Explorer/Assets/DCL/Interaction/Utility/RaycastUtils.cs:38]: It's a special type of raycasting originated from the center of the scene

File: Explorer/Assets/DCL/Interaction/PlayerOriginated/Systems/ProcessPointerEventsSystem.cs (hunk 1 above)
Comment (mikhail-dcl) [reply Explorer/Assets/DCL/Interaction/PlayerOriginated/Systems/ProcessPointerEventsSystem.cs:47]: Do you have ideas of what we can move to a separate method to improve readability **considerably**?

File: Explorer/Assets/Scripts/CrdtEcsBridge/Engine/EngineAPIImplementation.cs [hunk 4]
```diff
@@ -120,11 +120,15 @@ public ArraySegment<byte> CrdtSendToRenderer(ReadOnlyMemory<byte> dataMemory)
...
 
-            int payloadLength = SerializeOutgoingCRDTMessages();
-
             ApplySyncCommandBuffer(worldSyncBuffer);
 
-            return new ArraySegment<byte>(lastSerializationBuffer, 0, payloadLength);
+            if (returnData)
+            {
+                int payloadLength = SerializeOutgoingCRDTMessages();
+                return new ArraySegment<byte>(lastSerializationBuffer, 0, payloadLength);
```
[the remark this answers is not in the document]
Comment (mikhail-dcl) [reply Explorer/Assets/Scripts/CrdtEcsBridge/Engine/EngineAPIImplementation.cs:128]: it's perfectly fine as we don't need to propagate the exception across the whole managed and unmanaged internal of CLearScript. We just set an exception state in `exceptions handler` and suspend the scene

File: Explorer/Assets/Scripts/CrdtEcsBridge/Components/Special/ComponentInitializationExtensions.cs
[the remark this answers is not in the document]
Comment (mikhail-dcl) [reply Explorer/Assets/Scripts/CrdtEcsBridge/Components/Special/ComponentInitializationExtensions.cs:1]: done

**Review by m3taphysics — APPROVED** [review]

File: Explorer/Assets/DCL/Interaction/PlayerOriginated/Systems/ProcessPointerEventsSystem.cs (hunk 1 above)
Comment (popuz) [reply Explorer/Assets/DCL/Interaction/PlayerOriginated/Systems/ProcessPointerEventsSystem.cs:47]: Here an example of extraction to more high-level functions, so reader can skip if he is not needed to dive into low-level. Naming of the functions could be probably improved. Also, I am not sure auto-extraction properly used `ref` in all cases
```

[Query]
        private void ProcessRaycastResult(ref PlayerOriginRaycastResult raycastResult, ref HoverFeedbackComponent hoverFeedbackComponent,
            ref HoverStateComponent hoverStateComponent)
        {
            // Process all PBPointerEvents components to see if any of them is qualified
            hoverFeedbackComponent.Tooltips.Clear();

            bool candidateForHoverLeaveIsValid = TryGetPreviousEntityInfo(in hoverStateComponent, out GlobalColliderEntityInfo previousEntityInfo);
            hoverStateComponent.LastHitCollider = null;

            if (raycastResult.IsValidHit)
            {
                var entityInfo = raycastResult.EntityInfo.Value;

                // External world access should be always synchronized (Global World calls into Scene World)
                using (entityInfo.EcsExecutor.Sync.GetScope())
                {
                    candidateForHoverLeaveIsValid = 
                        ProcessGlobalToScene(raycastResult, ref hoverFeedbackComponent, ref hoverStateComponent, entityInfo.ColliderEntityInfo.EntityReference, 
                            entityInfo.EcsExecutor.World, candidateForHoverLeaveIsValid, previousEntityInfo);
                }
            }

            if (candidateForHoverLeaveIsValid)
                HoverFeedbackUtils.TryIssueLeaveHoverEventForPreviousEntity(in raycastResult, in previousEntityInfo);
        }

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        private bool ProcessGlobalToScene(PlayerOriginRaycastResult raycastResult,
            ref HoverFeedbackComponent hoverFeedbackComponent, ref HoverStateComponent hoverStateComponent,
            EntityReference entityRef, World world, bool candidateForHoverLeaveIsValid,
            GlobalColliderEntityInfo previousEntityInfo)
        {
            // Entity should be alive and contain PBPointerEvents component to be qualified
            if (!entityRef.IsAlive(world) || !world.TryGet(entityRef, out PBPointerEvents pbPointerEvents))
                return candidateForHoverLeaveIsValid;
            
            hoverStateComponent.LastHitCollider = raycastResult.UnityRaycastHit.collider;

            var newEntityWasHovered = !candidateForHoverLeaveIsValid
                                      || (previousEntityInfo.EcsExecutor.World != world &&
                                          previousEntityInfo.ColliderEntityInfo.EntityReference != entityRef);

            var anyInputInfo = sdkInputActionsMap.Values.GatherAnyInputInfo();
            pbPointerEvents.AppendPointerEventResultsIntent.Initialize(raycastResult.UnityRaycastHit, raycastResult.OriginRay);

            for (var i = 0; i < pbPointerEvents.PointerEvents.Count; i++)
                TryAppendPointerInfo(raycastResult, ref hoverFeedbackComponent, i, pbPointerEvents, anyInputInfo, newEntityWasHovered);

            // Signal to stop issuing hover leave event for the previous entity as it's equal to the current one
            if (candidateForHoverLeaveIsValid && !newEntityWasHovered)
                candidateForHoverLeaveIsValid = false;

            return candidateForHoverLeaveIsValid;
        }

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        private void TryAppendPointerInfo(PlayerOriginRaycastResult raycastResult,
            ref HoverFeedbackComponent hoverFeedbackComponent, int i, PBPointerEvents pbPointerEvents,
            InteractionInputUtils.AnyInputInfo anyInputInfo, bool newEntityWasHovered)
        {
            var pointerEvent = pbPointerEvents.PointerEvents[i];
            var info = pointerEvent.EventInfo;

            info.PrepareDefaultValues();

            if (!InteractionInputUtils.IsQualifiedByDistance(raycastResult, info)) return;

            // Check Input for validity
            InteractionInputUtils.TryAppendButtonLikeInput(sdkInputActionsMap, pointerEvent, i,
                ref pbPointerEvents.AppendPointerEventResultsIntent, anyInputInfo);

            // Try Append Hover Input
            if (newEntityWasHovered)
                InteractionInputUtils.TryAppendHoverInput(ref pbPointerEvents.AppendPointerEventResultsIntent,
                    PointerEventType.PetHoverEnter, pointerEvent, i);

            // Try Append Hover Feedback
            HoverFeedbackUtils.TryAppendHoverFeedback(sdkInputActionsMap, pointerEvent,
                ref hoverFeedbackComponent, anyInputInfo.AnyButtonIsPressed);
        }
```

**Pull request merged**